### PR TITLE
Enable servers to disable compression algorithms

### PR DESCRIPTION
**Pull request by dgquintas**
grpc-accept-encoding information is now propagated through the metadata

### CODE REVIEW

File: src/core/compression/algorithm.c [hunk 1]
```diff
@@ -37,11 +37,15 @@
...
 int grpc_compression_algorithm_parse(const char* name,
                                      grpc_compression_algorithm *algorithm) {
-  if (strcmp(name, "none") == 0) {
+  /* we use strncmp not only because it's safer (even though in this case it
+   * doesn't matter, given that we are comparing against string literals, but
+   * because this way we needn't have "name" nil-terminated (useful for slice
+   * data, for example) */
+  if (strncmp(name, "none", 4) == 0) {
     *algorithm = GRPC_COMPRESS_NONE;
-  } else if (strcmp(name, "gzip") == 0) {
```
Comment (ctiller) [inline src/core/compression/algorithm.c]: what happens when I introduce gzip2?

Comment (dgquintas) [inline src/core/compression/algorithm.c]: that a new awesome bug comes into existence, whereby behavior depends on the line ordering. Fixed. n00b bug shame cube.

Comment (ctiller) [inline src/core/compression/algorithm.c]: We should be able to add a test for that right?

Comment (dgquintas) [inline src/core/compression/algorithm.c]: We should. Done.

**Comment by dgquintas** [comment]
Merged with master. PTAL :)

**Comment by dgquintas** [comment]
Down to 9 changed files :)

File: src/core/channel/channel_args.c [hunk 2]
```diff
@@ -146,3 +147,45 @@ grpc_channel_args *grpc_channel_args_set_compression_algorithm(
   tmp.value.integer = algorithm;
   return grpc_channel_args_copy_and_add(a, &tmp, 1);
 }
+
+/** Returns the compression algorithm's enabled states bitset from \a a. If not
+ * found, return a biset will all algorithms enabled */
```
Comment (yang-g) [inline src/core/channel/channel_args.c]: bitset with?

Comment (dgquintas) [inline src/core/channel/channel_args.c]: indeed. fixed.

File: src/core/channel/channel_args.c [hunk 3]
```diff
@@ -146,3 +147,45 @@ grpc_channel_args *grpc_channel_args_set_compression_algorithm(
...
+  if (state != 0) {
+    GPR_BITSET(&states_bitset, algorithm);
+  } else {
+    GPR_BITCLEAR(&states_bitset, algorithm);
+  }
+
+  tmp.type = GRPC_ARG_INTEGER;
+  tmp.key = GRPC_COMPRESSION_ALGORITHM_STATE_ARG;
+  tmp.value.integer = states_bitset;
+  return grpc_channel_args_copy_and_add(a, &tmp, 1);
```
Comment (yang-g) [inline src/core/channel/channel_args.c]: Here `a` is copied and a new channel_args is malloced. Also there could be two states set in the resultant channel_args. Is this intended?

File: src/core/channel/channel_args.c [hunk 4]
```diff
@@ -146,3 +147,45 @@ grpc_channel_args *grpc_channel_args_set_compression_algorithm(
...
+}
+
+int grpc_channel_args_compression_algorithm_get_states(
+    const grpc_channel_args *a) {
+  return find_compression_algorithm_states_bitset(a);
```
Comment (yang-g) [inline src/core/channel/channel_args.c]: Do we want to use uint32 as the return value type? There is a uint32 to int conversion here.

File: src/cpp/server/server.cc [hunk 5]
```diff
@@ -185,28 +185,35 @@ class Server::SyncRequest GRPC_FINAL : public CompletionQueueTag {
   grpc_completion_queue* cq_;
 };
 
-static grpc_server* CreateServer(int max_message_size) {
+static grpc_server* CreateServer(
+    int max_message_size, const grpc_compression_options& compression_options) {
   if (max_message_size > 0) {
```
Comment (yang-g) [inline src/cpp/server/server.cc]: So you only care about the compression_options when there is a max_message_size set?

File: src/core/channel/channel_args.c (hunk 4 above)
Comment (yang-g) [inline src/core/channel/channel_args.c]: Never mind. It is an int in channel_args anyway.

File: src/cpp/server/server.cc (hunk 5 above)
Comment (dgquintas) [inline src/cpp/server/server.cc]: No, but CreateServer is a "private" (static) function that's always invoked with a max_message_size argument.

Comment (yang-g) [inline src/cpp/server/server.cc]: By default I think max_message_size is -1?

Comment (dgquintas) [inline src/cpp/server/server.cc]: Oh, I see now what you meant. Fixed.

**Comment by dgquintas** [comment]
Comments addressed and some bugs fixed. I've tested it manually over modified versions of the interop server/client. I'll extend the interop tests to exercise the accept-encoding metadata after this PR is merged.

PTAL

**Comment by dgquintas** [comment]
I've just changed the name of the PR to reflect its current contents. Over time, bits and pieces of the original one got merged. What remains is just the ability for servers to disable certain algorithms.

**Comment by dgquintas** [comment]
Ping. I've just merged against master's HEAD.

**Comment by yang-g** [comment]
Seems causing c test failures. Merge issue?

**Comment by dgquintas** [comment]
Yes. Fixed now. PTAL.

**Pull request merged**